.travis.yml

```diff
@@ -109,7 +112,7 @@ script:
   
 
   # run unit tests
-  - docker exec ${CONTAINER} /bin/sh -c "pip install pytest && cd /horovod/test && ${MPIRUN} pytest -v"
+  - docker exec ${CONTAINER} /bin/sh -c "pip install pytest && cd /horovod/test && ${MPIRUN} pytest -v --ignore=test_mxnet.py && ${MPIRUN} pytest -v test_mxnet.py"
```

Comment:
No order in which MXNet can co-exist with TensorFlow and PyTorch?

Reply:
We spent quite a bit time on this and let me try to explain what we found.

pytest loads all the tests by the alphabetic order of the filename. So the tests will run in the following order:
```
test_keras.py
test_mxnet.py
test_tensorflow.py
test_tensorflow_keras.py
test_torch.py
```
All the unit tests in each test file passed but we encountered an fatal error in MPI during pytest exit.
```
==================== 65 passed, 2 warnings in 12.94 seconds ====================

[c7949752ef7f:651] *** An error occurred in MPI_Gather
[c7949752ef7f:651] *** reported by process [2061107201,1]
[c7949752ef7f:651] *** on a NULL communicator
[c7949752ef7f:651] *** Unknown error
[c7949752ef7f:651] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[c7949752ef7f:651] ***    and potentially your MPI job)
[c7949752ef7f:651] [0] func:/usr/local/lib/libopen-pal.so.40(opal_backtrace_buffer+0x20) [0x7f6851c42090]
[c7949752ef7f:651] [1] func:/usr/local/lib/libmpi.so.40(ompi_mpi_abort+0x258) [0x7f68521cb4f8]
[c7949752ef7f:651] [2] func:/usr/local/lib/libmpi.so.40(ompi_mpi_errors_are_fatal_comm_handler+0xfd) [0x7f68521bbf0d]
[c7949752ef7f:651] [3] func:/usr/local/lib/libmpi.so.40(PMPI_Gather+0x269) [0x7f68521ec5a9]
[c7949752ef7f:651] [4] func:/usr/local/lib/python2.7/dist-packages/horovod/mxnet/mpi_lib.so(+0x2e050) [0x7f68391fd050]
[c7949752ef7f:651] [5] func:/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xb6990) [0x7f68ecd78990]
[c7949752ef7f:651] [6] func:/lib/x86_64-linux-gnu/libpthread.so.0(+0x8064) [0x7f68f4866064]
[c7949752ef7f:651] [7] func:/lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7f68f3c7862d]
```
I think there is an implicit requirement that we need to run all unit tests using the same backend framework together without interruption. Since Keras uses Tensorflow as backend, we need to run test_keras/test_tensorflow/test_tensorflow_keras together. With test_mxnet, it runs after test_keras and changes the order of running unit tests which causes the fatal error during exit.

It workes without test_mxnet because test_torch runs in the last without interrupting the tests using tensorflow backend. If we change test_torch to test_pytorch, the order of running unit tests is as follows:
```
test_keras.py
test_pytorch.py
test_tensorflow.py
test_tensorflow_keras.py
```
All the unit tests in each test file also passed but the same fatal error happened.
```
==================== 59 passed, 2 warnings in 11.53 seconds ====================
[c7949752ef7f:5213] *** An error occurred in MPI_Gather
[c7949752ef7f:5213] *** reported by process [1812398081,0]
[c7949752ef7f:5213] *** on a NULL communicator
[c7949752ef7f:5213] *** Unknown error
[c7949752ef7f:5213] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[c7949752ef7f:5213] ***    and potentially your MPI job)
[c7949752ef7f:5213] [0] func:/usr/local/lib/libopen-pal.so.40(opal_backtrace_buffer+0x20) [0x7f3be0698090]
[c7949752ef7f:5213] [1] func:/usr/local/lib/libmpi.so.40(ompi_mpi_abort+0x258) [0x7f3be0c214f8]
[c7949752ef7f:5213] [2] func:/usr/local/lib/libmpi.so.40(ompi_mpi_errors_are_fatal_comm_handler+0xfd) [0x7f3be0c11f0d]
[c7949752ef7f:5213] [3] func:/usr/local/lib/libmpi.so.40(PMPI_Gather+0x269) [0x7f3be0c425a9]
[c7949752ef7f:5213] [4] func:/usr/local/lib/python2.7/dist-packages/horovod/torch/mpi_lib_v2.so(+0x2f709) [0x7f3be00de709]
[c7949752ef7f:5213] [5] func:/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xb6990) [0x7f3c77508990]
[c7949752ef7f:5213] [6] func:/lib/x86_64-linux-gnu/libpthread.so.0(+0x8064) [0x7f3c8323c064]
[c7949752ef7f:5213] [7] func:/lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7f3c8264e62d]
```
We tried to rename test_mxnet to test_apache_mxnet which helped us get around this ordering issue locally because test_apache_mxnet run at the very beginning without interrupting the tests using tensorflow backend. However, it is flaky in Travis CI where we see unit test reordering happens (where test_keras started to run before we finished running all unit tests in test_apache_mxnet).

Running test_mxnet in a separate pytest session is the most stable way we found so far.

Reply:
OK, I think we can land this workaround for now until we figure out the root cause of NULL communicator issue - it does not seem to be related to MXNet.

Reply:
I'm testing https://github.com/uber/horovod/pull/730 which should fix this better.

Reply:
Great to know the root cause. Had the same feeling that we should run tests independently, at least for each framework.